File: webui/src/coordinates.ts
// Formulae adapted from wandergis/coordtransform (MIT).
// Copyright (c) 2015 记忆的残骸; see module/licenses/coordtransform.txt.
import { parsePosition, type Position } from './control';

export type CoordinateSystem = 'wgs84' | 'gcj02' | 'bd09';
type Point = { latitude: number; longitude: number };
const pi = Math.PI, xPi = pi * 3000 / 180, axis = 6378245, eccentricity = 0.00669342162296594323;
// Approximate conversion domain, not a geographic boundary dataset.
const inside = (p: Point) => p.longitude > 73.66 && p.longitude < 135.05 && p.latitude > 3.86 && p.latitude < 53.55;
function toGcj(p: Point): Point {
  if (!inside(p)) return p;
  const x = p.longitude - 105, y = p.latitude - 35;
  const waves = (20 * Math.sin(6 * x * pi) + 20 * Math.sin(2 * x * pi)) * 2 / 3;
  let lat = -100 + 2 * x + 3 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * Math.sqrt(Math.abs(x)) + waves;
  lat += (20 * Math.sin(y * pi) + 40 * Math.sin(y / 3 * pi)) * 2 / 3;
  lat += (160 * Math.sin(y / 12 * pi) + 320 * Math.sin(y * pi / 30)) * 2 / 3;
  let lon = 300 + x + 2 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * Math.sqrt(Math.abs(x)) + waves;
  lon += (20 * Math.sin(x * pi) + 40 * Math.sin(x / 3 * pi)) * 2 / 3;
  lon += (150 * Math.sin(x / 12 * pi) + 300 * Math.sin(x / 30 * pi)) * 2 / 3;
  const radians = p.latitude / 180 * pi;
  const magic = 1 - eccentricity * Math.sin(radians) ** 2, root = Math.sqrt(magic);
  return { latitude: p.latitude + lat * 180 / (axis * (1 - eccentricity) / (magic * root) * pi),
    longitude: p.longitude + lon * 180 / (axis / root * Math.cos(radians) * pi) };
}
function fromGcj(p: Point): Point {
  if (!inside(p)) return p;
  let result = { ...p };
  for (let i = 0; i < 20; i++) {
    const projected = toGcj(result);
    const latError = projected.latitude - p.latitude, lonError = projected.longitude - p.longitude;
    result = { latitude: result.latitude - latError, longitude: result.longitude - lonError };
    if (Math.max(Math.abs(latError), Math.abs(lonError)) < 1e-10) return result;
  }
  throw new Error('这个位置接近转换边界，请直接输入 WGS84 坐标');
}
function fromBd(p: Point): Point {
  const x = p.longitude - 0.0065, y = p.latitude - 0.006;
  const z = Math.hypot(x, y) - 0.00002 * Math.sin(y * xPi);
  const theta = Math.atan2(y, x) - 0.000003 * Math.cos(x * xPi);
  return { latitude: z * Math.sin(theta), longitude: z * Math.cos(theta) };
}
function toBd(p: Point): Point {
  const z = Math.hypot(p.longitude, p.latitude) + 0.00002 * Math.sin(p.latitude * xPi);
  const theta = Math.atan2(p.latitude, p.longitude) + 0.000003 * Math.cos(p.longitude * xPi);
  return { latitude: z * Math.sin(theta) + 0.006, longitude: z * Math.cos(theta) + 0.0065 };
}
export function convertCoordinates(position: Position, from: CoordinateSystem, to: CoordinateSystem): Position {
  parsePosition(String(position.latitude), String(position.longitude), String(position.altitude));
  if (from === to) return { ...position };
  const gcj = from === 'wgs84' ? toGcj(position) : from === 'bd09' ? fromBd(position) : position;
  const result = to === 'wgs84' ? fromGcj(gcj) : to === 'bd09' ? toBd(gcj) : gcj;
  parsePosition(String(result.latitude), String(result.longitude), String(position.altitude));
  return { ...position, ...result };
}

File: backend/src/coordinates.rs
// Formulae adapted from wandergis/coordtransform (MIT), also used by ui/src/coordinates.ts.
// Copyright (c) 2015 记忆的残骸; see backend/licenses/coordtransform.txt.
use crate::Position;
use serde::{Deserialize, Serialize};
use std::f64::consts::PI;

#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CoordinateSystem {
    Wgs84,
    Gcj02,
    Bd09,
}

fn inside(p: &Position) -> bool {
    p.longitude > 73.66 && p.longitude < 135.05 && p.latitude > 3.86 && p.latitude < 53.55
}
fn to_gcj(p: &Position) -> Position {
    if !inside(p) {
        return p.clone();
    }
    let (x, y) = (p.longitude - 105.0, p.latitude - 35.0);
    let waves = (20.0 * (6.0 * x * PI).sin() + 20.0 * (2.0 * x * PI).sin()) * 2.0 / 3.0;
    let mut lat =
        -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * x.abs().sqrt() + waves;
    lat += (20.0 * (y * PI).sin() + 40.0 * (y / 3.0 * PI).sin()) * 2.0 / 3.0;
    lat += (160.0 * (y / 12.0 * PI).sin() + 320.0 * (y * PI / 30.0).sin()) * 2.0 / 3.0;
    let mut lon = 300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * x.abs().sqrt() + waves;
    lon += (20.0 * (x * PI).sin() + 40.0 * (x / 3.0 * PI).sin()) * 2.0 / 3.0;
    lon += (150.0 * (x / 12.0 * PI).sin() + 300.0 * (x / 30.0 * PI).sin()) * 2.0 / 3.0;
    let radians = p.latitude.to_radians();
    let magic = 1.0 - 0.00669342162296594323 * radians.sin().powi(2);
    let root = magic.sqrt();
    let mut result = p.clone();
    result.latitude +=
        lat * 180.0 / ((6378245.0 * (1.0 - 0.00669342162296594323) / (magic * root)) * PI);
    result.longitude += lon * 180.0 / ((6378245.0 / root) * radians.cos() * PI);
    result
}
fn from_gcj(p: &Position) -> Result<Position, &'static str> {
    if !inside(p) {
        return Ok(p.clone());
    }
    let mut result = p.clone();
    for _ in 0..20 {
        let projected = to_gcj(&result);
        let (lat, lon) = (projected.latitude - p.latitude, projected.longitude - p.longitude);
        result.latitude -= lat;
        result.longitude -= lon;
        if lat.abs().max(lon.abs()) < 1e-10 {
            return Ok(result);
        }
    }
    Err("coordinate is near the conversion boundary; supply WGS84 directly")
}
pub fn convert(
    p: &Position,
    from: CoordinateSystem,
    to: CoordinateSystem,
) -> Result<Position, &'static str> {
    p.validate()?;
    if from == to {
        return Ok(p.clone());
    }
    let x_pi = PI * 3000.0 / 180.0;
    let gcj = match from {
        CoordinateSystem::Wgs84 => to_gcj(p),
        CoordinateSystem::Gcj02 => p.clone(),
        CoordinateSystem::Bd09 => {
            let (x, y) = (p.longitude - 0.0065, p.latitude - 0.006);
            let z = x.hypot(y) - 0.00002 * (y * x_pi).sin();
            let theta = y.atan2(x) - 0.000003 * (x * x_pi).cos();
            let mut result = p.clone();
            result.latitude = z * theta.sin();
            result.longitude = z * theta.cos();
            result
        }
    };
    let result = match to {
        CoordinateSystem::Wgs84 => from_gcj(&gcj)?,
        CoordinateSystem::Gcj02 => gcj,
        CoordinateSystem::Bd09 => {
            let z = gcj.longitude.hypot(gcj.latitude) + 0.00002 * (gcj.latitude * x_pi).sin();
            let theta = gcj.latitude.atan2(gcj.longitude) + 0.000003 * (gcj.longitude * x_pi).cos();
            let mut result = gcj.clone();
            result.latitude = z * theta.sin() + 0.006;
            result.longitude = z * theta.cos() + 0.0065;
            result
        }
    };
    result.validate()?;
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn conversions_round_trip_and_preserve_noncoordinate_fields() {
        let mut original = Position::new(31.2304, 121.4737);
        original.altitude = 42.0;
        for system in [CoordinateSystem::Gcj02, CoordinateSystem::Bd09] {
            let projected = convert(&original, CoordinateSystem::Wgs84, system).unwrap();
            let result = convert(&projected, system, CoordinateSystem::Wgs84).unwrap();
            assert!(crate::route::distance(&original, &result) < 0.15);
            assert_eq!(result.altitude, 42.0);
        }
        let overseas = Position::new(51.5, -0.1);
        assert_eq!(
            convert(&overseas, CoordinateSystem::Gcj02, CoordinateSystem::Wgs84).unwrap(),
            overseas
        );
    }
}
